sql: support timestamp done messages in TAIL

mjibson opened the pull request:
Add a `WITH (TIMESTAMPS)` option to TAIL that adds a `done` bool column
to the output. At the end of each set of batches send a new row with
the timestamp, a true `done`, and NULL for all columns to indicate
the timestamp is done and no other rows will appear at or before that
timestamp. Otherwise `done` is false.

This works because each batch represents a disjoint and increasing bound
of time. Thus we only need to send a done message for the last batch of
a group of batches.

Closes #4539

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/4752)
<!-- Reviewable:end -->

mjibson commented:
Last commit only, first is #4740. Open to better option name suggestions than `TIMESTAMPS` which is kind of bad because we already include timestamps and this adds done but `WITH (DONE)` is weird maybe?

frankmcsherry commented:
The concept elsewhere in timely is `PROGRESS`; maybe that would have less baggage?

ruchirK commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -66,6 +70,29 @@ pub fn tail<G>(
+                // results by timestamp so we can duplicate the last row's timestamp which is
+                // somewhat easier than using a batch.upper() antichain. This timestamp won't
+                // be reused any other batch because they are disjoint and strictly increasing.
+                let (time, _) = results.last().unwrap().clone();
```
there might still be multiple timestamps stored in results and if I'm reading this correctly, we would only emit a done message for the largest timestamp in `results` -- I'm not sure if thats the intent but if it is I think we should call that out in the docs more clearly?

mjibson replied:
You are correct. Added a paragraph documenting this.

frankmcsherry replied:
Another option here that communicates more information is to grab `batch.upper()` which will report the upper end of the open interval of timestamps contained in the batch. This doesn't have to be the largest timestamp in the batch; it could be larger.

ruchirK commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -66,6 +70,29 @@ pub fn tail<G>(
+                // by iterating through a row and skipping the first two that we add by
+                // hand. Remember this value and only iterate once because the iteration
+                // involves decoding the row datum bytes which is slow.
+                if cols == 0 {
```
I think if you were sinking a view with 2 columns this would do a lot more row decoding -- might be better to use an option for cols?

mjibson replied:
The `skip(2)` skips the `timestamp` and `done` columns. It should always be 1 (for `diff`) + (number of cols in the thing being tailed).

mjibson replied:
Added a comment to clarify.

ruchirK replied:
ah I see. I had glossed over the part where we add extra fields to the data rows. got it!

ruchirK commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -66,6 +70,29 @@ pub fn tail<G>(
+                // be reused any other batch because they are disjoint and strictly increasing.
+                let (time, _) = results.last().unwrap().clone();
+                packer.push(Datum::Decimal(Significand::new(i128::from(time))));
+                packer.push(Datum::True);
```
I don't know anything about the downstream protocol but it seems like the first two fields of this timestamp message / progress message could be totally different types than the rest of the sink -- is that just fine / handled nicely by COPY? asking mostly for my own edification

ruchirK replied:
ignore me! this is the same confusion as the last comment

ruchirK commented on src/dataflow-types/src/types.rs:
```diff
@@ -667,6 +667,7 @@ pub struct TailSinkConnector {
     pub tx: comm::mpsc::Sender<Vec<Row>>,
     pub frontier: Antichain<Timestamp>,
     pub strict: bool,
+    pub emit_timestamps: bool,
```
if you choose to rename to `PROGRESS` this variable would also be great to rename to `emit_progress`

mjibson commented:
I like `PROGRESS`. Any objections or other ideas?

ruchirK commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -66,6 +70,31 @@ pub fn tail<G>(
             // cursor will always produce rows in the same order.
             results.sort_by_key(|(time, _)| *time);
+
+            if connector.emit_timestamps && !batches.is_empty() {
```
you _might_ need to check that `results` is nonempty too? I am not 100% sure but i think if you tail without snapshot you can get nonempty batches that still dont emit anything into `results` which will cause a panic when you unwrap the last element

mjibson replied:
According to frank from slack:

> Because of timely dataflow constraints, DD cannot ship empty batches

so I think this is ok. Added a comment.

ruchirK replied:
right. I'm saying that even if the batch was not empty, you might not put anything in `results` because of as of / snapshot constraints. therefore I think you might still need to check that `results` is not empty

ruchirK replied:
e.g. i think if you got a batch covering timestamps [0, 4] but were tailing as of 5, `results` would be empty, which may cause a panic if you tried to emit progress messages. haven't tested this

frankmcsherry replied:
> Because of timely dataflow constraints, DD cannot ship empty batches

If at all possible it would be great to avoid relying on this; if DD could ship empty batches it would, and there are (not near-term) changes to timely that would allow this.

If the motivation is just to do this so that you can grab the time with confidence, I would consider using the batch's upper frontier as mentioned in a different comment!

ruchirK replied:
```
// Due to timely dataflow constraints, differential dataflow cannot ship empty
// batches, so we are guaranteed that if there's atleast one batch results will
// be non-empty.
```

I don't understand why the implication (if theres at least one batch then results will be nonempty) is true and if it is, could one of you explain why?

ruchirK commented:
I did a drive-by review on the dataflow code which all seems great to me! happy to review more if thats needed / wanted. thanks matt!

frankmcsherry reviewed (approved):
I have a few thoughts, which are mostly aimed at future proofing.

The more natural signal to provide at the end of a batch is the `upper` frontier, which in the case of a single timestamp is the *next* time at which we might see any records. This comes with each batch, so you can pull it out of even empty batches (not a thing we handle at the moment, but a thing that we should do eventually). It is larger than the last timestamp in the batch, so the meaning of `PROGRESS` would be announcing *next* times rather than "times you have seen that are now done".

This is all flexible, but I've implemented this same logic multiple times and each time I end up wanting the above.

benesch commented on doc/user/content/sql/tail.md:
```diff
@@ -75,6 +81,15 @@ By default, each TAIL is created with a `SNAPSHOT` which contains the results of
+
+If the `TIMESTAMPS` option is specified (`WITH (TIMESTAMPS)`) an additional `done` column appears in the output.
+It is `false` if there may be more rows with the same timestamp.
+It is `true` if there willbe no more rows at or before the timestamp, in which case all further columns will be `NULL`.
```
nit: missing space in "willbe"

mjibson commented:
Renamed option to `PROGRESS`, column renamed to `progressed`, batch upper now used instead of row time. The idea behind the column name `progressed` is that the timestamp has progressed to that time. I think that'll make sense to users? I've updated the docs to reflect. PTAL.

ruchirK reviewed (approved):
lgtm!

frankmcsherry commented on src/dataflow/src/sink/tail.rs:
```diff
@@ -76,3 +95,20 @@ pub fn tail<G>(
+fn batch_upper(
+    batch: Option<&Rc<OrdValBatch<GlobalId, Row, u64, isize, usize>>>,
+) -> Option<Timestamp> {
+    match batch {
+        Some(batch) => {
+            // Break the antichain abstraction here by fetching the first element because
+            // we know there won't ever be 2.
+            let upper = match batch.desc.upper().elements().get(0) {
+                Some(upper) => upper,
+                None => return None,
+            };
+            Some(*upper)
+        }
+        None => None,
+    }
```
I think this could possibly all be resolved down to `batch.map(|b| *b.desc.upper().elements.get(0))`, or is that wrong for some reason (Rust)?

mjibson replied:
Had to add `.flatten().map(..)` but yes thanks!

frankmcsherry reviewed (approved):
This looks good to me. Minor nit about possible Rust idiomatic code improvement, but it's also possible that the nit is wrong because I didn't try it myself!
